builder/codebuild/codebuild.go

```diff
@@ -0,0 +1,353 @@
+	DockerCfg string
+
+	// Function called to determine the CodeBuild project name, given a
+	// repository.
+	ProjectName func(builder.BuildOptions) string
+
+	codebuild codebuildClient
+
+	// This gets called when we need an io.Reader for a CloudWatch log
+	// stream.
+	logReader func(group, stream string) (io.ReadCloser, error)
+}
+
+func NewBuilder(config client.ConfigProvider) *Builder {
+	cloudwatchlogs := cloudwatchlogs.New(config)
+
+	// Use the cloudwatch package to get an io.Reader for a cloudwatch log
+	// stream.
+	logReader := func(group, stream string) (io.ReadCloser, error) {
+		return cloudwatch.NewReader(group, stream, cloudwatchlogs), nil
+	}
+
+	return &Builder{
+		BuildSpec:   template.Must(template.New("buildspec.yml").Parse(DefaultBuildspec)),
+		ProjectName: ProjectName,
+		DockerCfg:   "conveyor.dockercfg",
+		codebuild:   codebuild.New(config),
+		logReader:   logReader,
+	}
+}
+
+// Build executes the docker image.
+func (b *Builder) Build(ctx context.Context, w io.Writer, opts builder.BuildOptions) (image string, err error) {
+	projectName := b.ProjectName(opts)
+	image = fmt.Sprintf("%s:%s", opts.Repository, opts.Sha)
+	err = b.build(ctx, w, BuildOptions{
+		BuildOptions: opts,
+		ProjectName:  projectName,
+		Dry:          b.Dry,
+		DockerCfg:    b.DockerCfg,
+	})
+	return
+}
+
+func (b *Builder) build(ctx context.Context, w io.Writer, opts BuildOptions) error {
+	resp, err := b.startBuild(ctx, opts)
+	if err != nil {
+		// if the project does not exist, create it.
+		if awsErr, ok := err.(awserr.Error); ok && awsErr.Code() == "ResourceNotFoundException" {
+			if _, err := b.createProject(ctx, opts); err != nil {
+				return err
+			}
+
+			resp, err = b.startBuild(ctx, opts)
+			if err != nil {
+				return err
+			}
+		} else {
+			return fmt.Errorf("starting build for %s: %v", opts.Repository, err)
+		}
+	}
+
+	buildID := resp.Build.Id
+
+	done := make(chan error)
+	go func() {
+		done <- b.tailBuild(ctx, w, buildID)
+	}()
+
+	select {
+	case <-ctx.Done():
+		// Try cancelling the build. We don't really care about the
+		// error here.
+		b.codebuild.StopBuild(&codebuild.StopBuildInput{
+			Id: buildID,
+		})
+
+		// No wait for log streaming and everything to finish up.
+		err = <-done
+		if err == nil {
+			err = ctx.Err()
+		}
+	case err = <-done:
+	}
+
+	if err == context.Canceled || err == context.DeadlineExceeded {
+		return &builder.BuildCanceledError{
+			Err:    err,
+			Reason: ctx.Err(),
+		}
+	}
+
+	return err
+}
+
+func (b *Builder) tailBuild(ctx context.Context, w io.Writer, buildID *string) error {
+	fmt.Fprintf(w, "conveyor: Waiting for CodeBuild build %s to start\n", *buildID)
+
+	build, err := b.waitUntilLogsAvailable(ctx, buildID)
+	if err != nil {
+		return err
+	}
+
+	fmt.Fprintf(w, "conveyor: CodeBuild build %s started, streaming logs...\n", *buildID)
+
+	r, err := b.logReader(*build.Logs.GroupName, *build.Logs.StreamName)
+	if err != nil {
```

Comment:
Annoyingly, the `LogsLocation` isn't available immediately after starting a build, so you have to poll until it's available, this is handled on the [line above]( https://github.com/remind101/conveyor/pull/72/files/b71b1a45360308aabc50428d2cf63a9b581a10be#diff-8fb7f6f7e9814f425fe9e649d7436038R195). Shouldn't be any chance for a race here.